### ISSUE TITLE
Provide support for type aliases

### ISSUE THREAD

**Issue by mitar** [protocolbuffers/protobuf#3521]
So that we could define maybe something like:

```
type dataset_uri = string;

message StartTrain {
  dataset_uri from = 1;
  dataset_uri to = 2;
}
```

This would allow much more readable proto files.

**Comment by pherl** [comment]
The main concern would be increase of complexity.  Also I doubt it will increase readability, as users now would have to look up the definition of the types to figure out the real type for the generated API - you cannot distinguish between an aliased type vs a message type. I guess the benefit doesn't outweigh the cost, and comments would probably be better here for documentation purposes.

**Comment by david-aimsun** [comment]
The complexity can be handle by tools. For example a plugin for Visual Code can show extra info: original type and documentation. IMHO, will make the code more readable. I prefer types (I´m a C++ guy) than strings everywhere (or, in our case, uint64 representing milliseconds since epoch).

**Comment by YamiOdymel** [comment]
I'd love to see this got implemented. The type alias tells more information rather than a simple `uint64` or `string` type. 

And finally, the developer can use type alias to describe how the _real_ API is. With the Protobuf built-in types, it will never be enough and that's what really confuses us.

---

Also, like _david-aimsun_ says, it can be handle by tools like this.

![image](https://user-images.githubusercontent.com/7308718/51660738-afdf5180-1fe9-11e9-8e4b-8eb1023bfcb3.png)

**Comment by ramosbugs** [comment]
It's also potentially useful to be able to associate [field options](https://developers.google.com/protocol-buffers/docs/proto#options) (including custom ones) with a type alias to avoid having to duplicate them in many places.

**Comment by micheal-hill** [comment]
I know this is an old issue, but this would be handy for me too. I have a definition that contains a URI field, generated token and a user-provided text field; right now I can only type these each as `string`. This is being used for an API between teams though; so I need to document elsewhere that each of these fields must conform to some extra constraints.

**Comment by awkr** [comment]
I'd like to have this feature, cause it'll improve the abstract and more object-oriented. Such as there is a CURD API with most fields are `repeated string`, with this feature, I can only type `strings` or `strs`, somehow elegant.

**Comment by cbentzel** [comment]
I was about to file a similar issue. Right now if there is a repeated use of a scalar type across messages (say, a uint32 to represent an identifier) there are two ways to help ensure consistency:

- Nested messages, but this comes with run-time overhead.
- Commenting and careful review to ensure that the types are the same.

If one ever needs to increase the size of the scalar - say from uint32 to uint64  - there is no way to do this without manually scrubbing the files. Using a typedef/alias would allow substitution in one place.

This could also be done by having protobuf users write custom pre-processors which does the type alias substitution before sending to protoc, but seems like something that would be useful in the proto syntax and protoc itself.

**Comment by Fleshgrinder** [comment]
That's actually how I would implement it in Protobuf. Have a pre-process phase that replaces all type aliases with their actualy definitions and add an option to them that there was a type alias. Afterwards send it to the actual code generators. The overall impact on the Protobuf ecosystem would be minimal because all code generators that are currently in existence would continue to work since they already have to support custom options anyways. Code generators that are interested in the type aliases (e.g. `typealias Uuid = string` is mapped to `java.util.UUID` in the generated code`) can do so by inspecting the custom e.g. `google.protobuf.typealias` option‘s value.

**Comment by mattdornfeld** [comment]
Thrift does support this in the form of typedefs. The use case of this is clearer communication of what the type should be used for and better type safety in the code that imports the generated protobuf code.

**Comment by mjfryc** [comment]
I'm using [protodot](https://github.com/seamia/protodot) to generate structure graph in .dot format and [protoc-gen-doc](https://github.com/pseudomuto/protoc-gen-doc) for documentation. It will be very useful to visualize alias types dependencies at the graph and alias-type-links inside md/html/pdf documentation.

Currently the only hack is:

    message my_entity_id_type {
        string id = 1; // I would like any `typealias my_entity_id_type = string;` instead
    }

**Comment by chandramouleswaran** [comment]
+1 on the ask - not sure why this is closed. In C#, we would like to convert sensitive strings to SecureStrings after deserialization.
Bond supports this really well using type alias [converters](https://microsoft.github.io/bond/manual/bond_cs.html#converter)

**Comment by chandramouleswaran** [comment]
+1 - Not sure why this got closed - alias is really useful for cases where we cannot keep primitive data types in memory or persisted. In C#, we would like an [alias converting](https://microsoft.github.io/bond/manual/bond_cs.html#converter) mechanism similar to Bond. 

Cases where we can convert string to/from secure string etc.

**Comment by fowles** [comment]
This is something we intend to look into although it likely will not happen in 2021

**Comment by fowles** [comment]
For those curious why this is not a very simply thing to add.  aliases for messages interacts with parsing of Any fields and so it needs to be designed carefully to ensure long term stability.

**Comment by chandramouleswaran** [comment]
> This is something we intend to look into although it likely will not happen in 2021

Is this planned for 2022? Any specific version to follow?

**Comment by perezd** [comment]
This is not currently planned for 2022.

**Comment by Fleshgrinder** [comment]
> For those curious why this is not a very simply thing to add. aliases for messages interacts with parsing of Any fields and so it needs to be designed carefully to ensure long term stability.

@fowles how come? A type alias should not have any effect on `Any` since it still is the original type. Adding support for type aliases really should just be a pre-processing step that resolves each level of aliasing down to the original type while merging metadata together as needed to finally replace any usage with the actual type and all the metadata.

```protobuf
package example;

type Uuid = string [ (validate.rules).string = { uuid: true } ];

type A = string [ (custom).x = true, (custom).y = true ];

type B = A [ (custom).x = false, (custom).z = true ];

message M {
  Uuid id = 1;
  A a = 2;
  B b = 3;
}
```

We turn this into before we hand it out to any protoc plugin:

```protobuf
package example;

message M {
  string id = 1 [
    alias = "example.Uuid",
    (validate.rules).string = { uuid: true }
  ];

  string a = 2 [
    alias = "example.A",
    (custom).x = true,
    (custom).y = true
  ];

  string b = 3 [
    alias = "example.B",
    (custom).x = false,
    (custom).y = true,
    (custom).z = true
  ];
}
```

Basically the alias is gone entirely. Plugins that care about them can look at the metadata and react to it on a per field basis. The aliases themselves are not reflectible or accessible in any way. This really covers like 99% of all use cases. The only thing that is not possible is that a protoc plugin alters an alias, but I would not see this as a blocker to achieve a minimally invasive implementation that can be rolled out without fear.

**Comment by haberman** [comment]
"Type alias" is a somewhat overloaded term; the feature we have contemplated internally is a *message* type alias:

```protobuf
syntax = "proto3";

message Foo {}

// Hypothetical syntax.
//
// This would actually generate a type alias in each target language,
// so that both names are considered equivalent. 
message Bar = Foo;

message Baz {
  Bar bar = 1;
}
```

The motivation of message aliases would be to facilitate renaming of messages while allow clients to gradually migrate from the old name to the new name, very similar to the rationale given in [this Go proposal](https://go.googlesource.com/proposal/+/master/design/18130-type-alias.md):

> The primary motivation is to enable gradual code repair during large-scale refactorings, in particular moving a type from one package to another in such a way that code referring to the old name interoperates with code referring to the new name. 

Message aliases do end up causing problems for `Any`, since message names are embedded into wire payloads for `Any` messages, so changing the message name is a wire-breaking change.

The version of type aliases given on this thread, where the alias is a mere substitution of the field type (and custom options), would not have any problems with `Any`, I agree. But it also has not been as frequent of a feature request, and therefore has a less clear benefit to the relatively large amount of complexity it would introduce.

**Comment by Fleshgrinder** [comment]
That is indeed going to interfere with `Any` and I see where you're coming from with this. I also agree that this would be very useful. However, it's not what this issue is or was about. Making it about an entirely different thing is not a good idea as it's only going to lead to disappointment. The fact that we have an issue for aliases for any kind of symbol that can act as a lightweight replacement to consistently copy around semantics and metadata but that we don't have an issue for the message aliasing as you described it proves that there is a need and high interest.

I've never encountered an aliasing system in any language that actually creates new types. Go would be the first I know off if that proposal is accepted. At the same time the naming does make sense, however, it seems to be chosen because `type` declarations already exist.

I clearly see merit in both features and there should be a tracking issue for what you described.

**Comment by KaymeKaydex** [comment]
An alias is something that will allow me not to rewrite the same lines a million times and replace plugin descriptions in these lines instantly everywhere and at once

    string Email = 1 [
        json_name = "email",
        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
            description: "User email"
            example: "testuser@example.ru"
        },
        (validate.rules).string = {
            email: true
            max_len: 1024
            ignore_empty: false
        }
    ];

and the attached message will be a nested field, for example if I use "message email" it looks like email.email

**Comment by liu-xuewen** [comment]
curdl require a message alias to avoid duplicate message definitions

**Comment by panagiotidisc** [comment]
I'm mapping a predetermined list of strings from an external api to my own enum.  I need to know the original string representation, and unfortunately some of the strings use characters that cannot be used in the name of the enum value (eg "foo-bar"). As such I have created a custom option to put the original name, and because it's widely applicable I have put it in its own file to share with other protos. The issue is that every file that imports it must use the fully qualified name, instead of something more DRY.
I want to turn every instance of this:
`[(MyEnumOptions.enum_name) = "foo"]`
to this
`[(enum_name) = "foo"]`
to make it similar to the much more concise json_name option.

**Comment by zyxkad** [comment]
My personal issue is that I don't like to write a really long type name again and again when import another proto file, (e.g. `google.protobuf.Any`, `google.protobuf.Empty`)
I wish type alias can be add soon. I believe nobody would like write something like `java.lang.String = xxx` in java

**Comment by noorus** [comment]
I think the case for this is incredibly obvious. And I mean basic type aliases in particular, not message aliasing, which probably CAN be harder to implement.

Example:
We have a project which communicates timestamps. The timestamps have been decided to be be uint64:s.
I will write a ton of messages with fields like:
```
message Packet_Foobar {
  uint64 time = 1;
}
```

Now, if we EVER decide that uint64 is not the ideal underlying basic type for timestamps, instead of replacing one aliasing like like `alias Timestamp = uint64;`, I am going to have to:

**manually find and replace every single place where we have ever communicated a timestamp inside a message and change the data type by hand, almost certainly missing one or more places or making other errors.**

How on earth is this:
- Better or more secure than just having a single-line "using"-like type alias
- Hard to implement (it's literally just a `protoc` preprocessor step)
- Not useful
- Not on the roadmap

It does not compute, it just doesn't

**Comment by zyxkad** [comment]
ping

**Comment by sbussard** [comment]
bump

**Comment by haberman** [comment]
This is not something we've identified as a pain point at Google, and so I think it's unlikely that we'd prioritize this.

Beyond that, I think this feature might be dangerous as proposed. Changing the type of a field is both an API-breaking change and (usually) a wire-breaking change too. For example, changing a field from `uint64` to `string` will definitely break both users of the generated API and break the wire representation of that message.

Suppose you have a line like:

```protobuf
alias Timestamp = uint64;
```

If you change this, it will break both API and wire for N other messages. There's no way of knowing how many messages will be affected by looking at the type alias alone. If the type alias can be used through imports, it could even break the messages of your users.

If you have a common type that you might want to change later, the idiomatic way of representing this with Protocol Buffers is to define a message:

```protobuf
message Timestamp {
  // ...
}
```

This adds a bit more overhead, but has the benefit of being possible to extend or change without breaking API/wire compatibility.

**Comment by vphantom** [comment]
Shouldn't this be closed as `wontfix` instead of `completed`?

**Comment by haberman** [comment]
Yes, good point. Done.

**Comment by FreeSpirit9009** [comment]
Dozens of users _have_ identified it as a pain point, I do find it a bit strange that Google seems to know better than it's users. But Google is Google...
